### PR TITLE
Updates the ttl-to-finished KEP to graduate the feature to Beta.

### PR DESCRIPTION
**Pull request by ahg-g**
Enhancement issue: #592

### CODE REVIEW

**Comment by ahg-g** [comment]
@janetkuo should I move the "Finished Pods" section to a separate "Future Work" section?

File: keps/sig-apps/592-ttl-after-finish/README.md
Comment (janetkuo) [inline keps/sig-apps/592-ttl-after-finish/README.md:262]: Graduation criteria should be based on quality/maturity instead of releases. This can be captured in the `milestone` of kep.yaml.

File: keps/sig-apps/592-ttl-after-finish/README.md
Comment (wojtek-t) [inline keps/sig-apps/592-ttl-after-finish/README.md:261]: Can you describe the tests that should be added?

File: keps/sig-apps/592-ttl-after-finish/README.md
Comment (wojtek-t) [reply keps/sig-apps/592-ttl-after-finish/README.md:262]: Agree with Janet. Also please split by alpha-beta and beta->ga graduation, see e.g. here:
https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/1904-efficient-watch-resumption#graduation-criteria

File: keps/sig-apps/592-ttl-after-finish/README.md
Comment (wojtek-t) [inline keps/sig-apps/592-ttl-after-finish/README.md:288]: With caveats (as this is API field).

Please take a look at https://github.com/kubernetes/enhancements/pull/2199/files#diff-b18c53698772d9f0dcc15e9f35d90ecd0f181923f33e40115036991aa0b5f1c5R234 - it's a pretty similar case wrt to this question.

File: keps/sig-apps/592-ttl-after-finish/README.md
Comment (wojtek-t) [inline keps/sig-apps/592-ttl-after-finish/README.md:300]: nit: I think it's silently dropped if the feature is disabled (which is slightly different than accepted if enabled)

File: keps/sig-apps/592-ttl-after-finish/README.md
Comment (wojtek-t) [inline keps/sig-apps/592-ttl-after-finish/README.md:304]: How about kube-apiserver?
I suggest also extended 4xx/5xx on Jobs from kube-apiserver (similarly to what I've been pointing out in #2199 that I mentioned above).

File: keps/sig-apps/592-ttl-after-finish/README.md
Comment (wojtek-t) [inline keps/sig-apps/592-ttl-after-finish/README.md:307]: Definitely not N/A - there are ways that it can potentially break.

File: keps/sig-apps/592-ttl-after-finish/README.md
Comment (wojtek-t) [inline keps/sig-apps/592-ttl-after-finish/README.md:318]: How you can get that from workqueue metrics?

File: keps/sig-apps/592-ttl-after-finish/README.md
Comment (wojtek-t) [inline keps/sig-apps/592-ttl-after-finish/README.md:319]: nit: it's just a wording, but I'm trying to introduce a bit of consistency between answers (we will hopefully get to to more "choose from the below options" in the future here):

`Listing jobs in the cluster and checking if any has ttlSecondsAfterFinished field set.`

File: keps/sig-apps/592-ttl-after-finish/README.md
Comment (wojtek-t) [inline keps/sig-apps/592-ttl-after-finish/README.md:336]: I don't see an easy way to compute such SLO out of the existing metrics.

Can we create a metric that will basically expose this information?
Something like:
"histogram of when the job were supposed to be deleted minus when it was deleted"
[somewhat similar to what we did (or supposed to do) for cronjob: https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/19-Graduate-CronJob-to-Stable#metrics ]

File: keps/sig-apps/592-ttl-after-finish/README.md
Comment (wojtek-t) [inline keps/sig-apps/592-ttl-after-finish/README.md:354]: maybe: "at most one for each job creation? (depending on feature use)"

[that basically puts the upper bound and shows that while we don't control the throughput per-se, the total amount of requests is bounded by requests that already exist]

File: keps/sig-apps/592-ttl-after-finish/README.md
Comment (wojtek-t) [inline keps/sig-apps/592-ttl-after-finish/README.md:366]: Technically yes - we're adding a field so the object will be slightly bigger :)

File: keps/sig-apps/592-ttl-after-finish/kep.yaml
Comment (wojtek-t) [inline keps/sig-apps/592-ttl-after-finish/kep.yaml:8]: implementable

[status should be changed to implemented after GA graduation]

File: keps/sig-apps/592-ttl-after-finish/README.md
Comment (ahg-g) [reply keps/sig-apps/592-ttl-after-finish/README.md]: updated to indicate that the field will continue to exist but should have no effect.

File: keps/sig-apps/592-ttl-after-finish/README.md
Comment (ahg-g) [reply keps/sig-apps/592-ttl-after-finish/README.md]: Updated.

File: keps/sig-apps/592-ttl-after-finish/README.md
Comment (ahg-g) [reply keps/sig-apps/592-ttl-after-finish/README.md]: done.

File: keps/sig-apps/592-ttl-after-finish/README.md
Comment (ahg-g) [reply keps/sig-apps/592-ttl-after-finish/README.md]: updated the sentence

File: keps/sig-apps/592-ttl-after-finish/README.md
Comment (ahg-g) [reply keps/sig-apps/592-ttl-after-finish/README.md]: done.

File: keps/sig-apps/592-ttl-after-finish/README.md
Comment (ahg-g) [reply keps/sig-apps/592-ttl-after-finish/README.md]: updated

File: keps/sig-apps/592-ttl-after-finish/README.md
Comment (ahg-g) [reply keps/sig-apps/592-ttl-after-finish/README.md]: updated

File: keps/sig-apps/592-ttl-after-finish/README.md
Comment (ahg-g) [reply keps/sig-apps/592-ttl-after-finish/README.md:276]: is that a metric that already exists? if so, what is the name?

File: keps/sig-apps/592-ttl-after-finish/README.md
Comment (ahg-g) [reply keps/sig-apps/592-ttl-after-finish/README.md]: updated to manually tested as part of alpha to beta graduation criteria

File: keps/sig-apps/592-ttl-after-finish/README.md
Comment (ahg-g) [reply keps/sig-apps/592-ttl-after-finish/README.md:307]: I was thinking we could use `workqueue_queue_duration_seconds`: https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/component-base/metrics/prometheus/workqueue/metrics.go#L60

File: keps/sig-apps/592-ttl-after-finish/kep.yaml
Comment (ahg-g) [reply keps/sig-apps/592-ttl-after-finish/kep.yaml]: done

File: keps/sig-apps/592-ttl-after-finish/README.md
Comment (ahg-g) [reply keps/sig-apps/592-ttl-after-finish/README.md]: good point.

File: keps/sig-apps/592-ttl-after-finish/README.md
Comment (wojtek-t) [reply keps/sig-apps/592-ttl-after-finish/README.md:277]: apiserver_request_total has a "code" (being http code returned to the user):
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go#L76

File: keps/sig-apps/592-ttl-after-finish/README.md
Comment (wojtek-t) [inline keps/sig-apps/592-ttl-after-finish/README.md:291]: I think you need to provide some label to that metric (controller name or sth like that), right?
Can you please make that explicit?

File: keps/sig-apps/592-ttl-after-finish/README.md
Comment (wojtek-t) [inline keps/sig-apps/592-ttl-after-finish/README.md:298]: Same here (request for the label name)

File: keps/sig-apps/592-ttl-after-finish/README.md
Comment (wojtek-t) [reply keps/sig-apps/592-ttl-after-finish/README.md:308]: Can you describe how? [I think you will be putting all updates to Job objects in the queue, so distinguishing the ones that are relevant might be hard in that metric.]

File: keps/sig-apps/592-ttl-after-finish/README.md
Comment (ahg-g) [reply keps/sig-apps/592-ttl-after-finish/README.md:277]: you mean 4xx/5xx replies when reaching to the jobs endpoint on the api-server? why would that be relevant to this feature?

Comment (wojtek-t) [reply keps/sig-apps/592-ttl-after-finish/README.md:277]: They don't have to be relevant. But they may potentially be, e.g. imagine tha tthe conversion function is returning an error if that field is set - that would exactly result in returning 5xx from it.

File: keps/sig-apps/592-ttl-after-finish/README.md
Comment (wojtek-t) [inline keps/sig-apps/592-ttl-after-finish/README.md:280]: Any findings?

File: keps/sig-apps/592-ttl-after-finish/README.md
Comment (wojtek-t) [reply keps/sig-apps/592-ttl-after-finish/README.md:309]: Thanks for explanation.

That said - I don't think you can use workqueue metrics to report what we want exactly:
1) We really want to wait for TTL - we don't want to delete jobs too early
2) The fact that the queue is keeping up with load doesn't mean there isn't a bug in the code and we aren't really deleting as fast as we should.

So I really think we should add a metric similar to what was done for CronJob...

Comment (ahg-g) [reply keps/sig-apps/592-ttl-after-finish/README.md:309]: > We really want to wait for TTL - we don't want to delete jobs too early

We are not deleting jobs too early. Not sure how you concluded that from my comment.

> The fact that the queue is keeping up with load doesn't mean there isn't a bug in the code and we aren't really deleting as fast as we should.

Where would the metric be updated? To achieve the semantics you are looking for I think we need a dedicated metric endpoint such that when queried, all jobs are listed and a histogram is computed (something like https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/1748-pod-resource-metrics).

File: keps/sig-apps/592-ttl-after-finish/README.md
Comment (ahg-g) [reply keps/sig-apps/592-ttl-after-finish/README.md:280]: oh, sorry, I thought I would be doing that before submitting the flag flip PR, I will do the test today.

**Review by janetkuo — APPROVED** [review]
LGTM. I'll let @wojtek-t do the PRR review.

File: keps/sig-apps/592-ttl-after-finish/README.md
Comment (ahg-g) [reply keps/sig-apps/592-ttl-after-finish/README.md]: done.

File: keps/sig-apps/592-ttl-after-finish/README.md
Comment (ahg-g) [reply keps/sig-apps/592-ttl-after-finish/README.md:281]: I had some issues with gce-enforcer (deleting firewall rules from my k8s on gce cluster), will try again tomorrow.

Comment (ahg-g) [reply keps/sig-apps/592-ttl-after-finish/README.md:281]: done, works as expected.

File: keps/sig-apps/592-ttl-after-finish/README.md
Comment (wojtek-t) [inline keps/sig-apps/592-ttl-after-finish/README.md:274]: The field isn't dropped, right?
We're preserving the fields if they were set.

File: keps/sig-apps/592-ttl-after-finish/README.md
Comment (wojtek-t) [reply keps/sig-apps/592-ttl-after-finish/README.md:310]: > We are not deleting jobs too early. Not sure how you concluded that from my comment.

Sorry - I didn't mean we're deleting.
What I was saying: "what if there is a bug that leads to deleting them to early".

In general, the purpose of many of metrics isn't needed is everything works as designed. They are used if there are bugs, something doens't work as expected, etc....

> Where would the metric be updated? To achieve the semantics you are looking for I think we need a dedicated metric endpoint such that when queried, all jobs are listed and a histogram is computed (something like https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/1748-pod-resource-metrics).

What I had on mind is basically:
- when you send a "DELETE job" request, you basically export the metric  "currentTime - time when job succeeded - TTL value"

Sure - if the feature doesn't work at all we won't see anything in metrics. But assuming it eventually deletes stuff, we will see how accurately it does.

File: keps/sig-apps/592-ttl-after-finish/README.md
Comment (ahg-g) [reply keps/sig-apps/592-ttl-after-finish/README.md:274]: It is dropped: https://github.com/kubernetes/kubernetes/blob/d72c056260e771e8cd01220203582de0c0015786/pkg/registry/batch/job/strategy.go#L79

Comment (wojtek-t) [reply keps/sig-apps/592-ttl-after-finish/README.md:274]: Hmm - which isn't really a pattern we want to have. We generally want to avoid clearing explicitly set fields, e.g. in:
https://github.com/kubernetes/kubernetes/blob/d72c056260e771e8cd01220203582de0c0015786/pkg/api/pod/util.go#L450

@liggitt - FYI

File: keps/sig-apps/592-ttl-after-finish/README.md
Comment (ahg-g) [reply keps/sig-apps/592-ttl-after-finish/README.md:310]: > In general, the purpose of many of metrics isn't needed is everything works as designed. They are used if there are bugs, something doens't work as expected, etc....

Right, but my general argument is that adding a new metric isn't going to make things a lot different compared to relying on the workqueue ones with respect to avoiding bugs.

 > when you send a "DELETE job" request, you basically export the metric "currentTime - time when job succeeded - TTL value"
> Sure - if the feature doesn't work at all we won't see anything in metrics. But assuming it eventually deletes stuff, we will see how accurately it does.

This doesn't protect us from bugs in the workqueue (your initial concern about the workqueue metrics), what if the workqueue has a bug were push sometimes work and sometimes silently doesn't (and so some jobs will get deleted and some will not or randomly delayed)? 

In any case, I updated the KEP to indicate that we will add this metric. But, our source of truth is etcd, and I think the only way to reliably achieve the semantics you want is by querying that directly.

File: keps/sig-apps/592-ttl-after-finish/README.md
Comment (ahg-g) [reply keps/sig-apps/592-ttl-after-finish/README.md:281]: done.

File: keps/sig-apps/592-ttl-after-finish/README.md
Comment (ahg-g) [reply keps/sig-apps/592-ttl-after-finish/README.md:274]: This is an existing behavior, do we want to change it?

Comment (wojtek-t) [reply keps/sig-apps/592-ttl-after-finish/README.md:274]: I think we want - though please reach out to @liggitt (or other API approver) for final confirmation.

File: keps/sig-apps/592-ttl-after-finish/README.md
Comment (wojtek-t) [reply keps/sig-apps/592-ttl-after-finish/README.md:315]: I agree that doesn't protect us from all kind of bugs, but I think that metric itself is fairly useful on its own.

Anyway - thanks!

File: keps/sig-apps/592-ttl-after-finish/kep.yaml
Comment (wojtek-t) [inline keps/sig-apps/592-ttl-after-finish/kep.yaml:49]: Please also add ttl_after_finished_controller_time_to_deletion_seconds metric here

**Review by wojtek-t — COMMENTED** [review]
I'm fine with it modulo one minor nit and the open comment about dropping field - please resolve this one with API approvers.

You also need sig-apps approval.

**Comment by wojtek-t** [comment]
> You also need sig-apps approval.

Actually - I see Janet already approved. So I will just wait for the "drop fields" discussion resolution and will approve.

File: keps/sig-apps/592-ttl-after-finish/README.md
Comment (liggitt) [reply keps/sig-apps/592-ttl-after-finish/README.md:274]: The current behavior looks correct.. if the feature is disabled, we only preserve the field if it was already set in the **persisted** object, and clear it otherwise. That

Comment (ahg-g) [reply keps/sig-apps/592-ttl-after-finish/README.md:274]: done.

File: keps/sig-apps/592-ttl-after-finish/kep.yaml
Comment (ahg-g) [reply keps/sig-apps/592-ttl-after-finish/kep.yaml:49]: done.

**Comment by wojtek-t** [comment]
LGTM

@ahg-g - please squash commits and I will approve

**Comment by ahg-g** [comment]
> LGTM
> 
> @ahg-g - please squash commits and I will approve

thanks for the review, done!

File: keps/sig-apps/592-ttl-after-finish/README.md
Comment (wojtek-t) [inline keps/sig-apps/592-ttl-after-finish/README.md:260]: Your rebase didn't work - it seems you removed all the further commits....

Comment (wojtek-t) [reply keps/sig-apps/592-ttl-after-finish/README.md:260]: :(

Comment (ahg-g) [reply keps/sig-apps/592-ttl-after-finish/README.md]: ops, forgot to `git add .` :) should be fixed now

**Comment by wojtek-t** [comment]
Thanks!

/lgtm
/approve

**Pull request merged**
